Source/ParticleInterpolators/ParticleInterpolators.impl.hpp

```diff
@@ -0,0 +1,655 @@
+/* GRTeclyn
+ * Copyright 2022 The GRTL collaboration.
+ * Please refer to LICENSE in GRTeclyn's root directory.
+ */
+
+#if !defined(PARTICLEINTERPOLATORS_HPP_)
+#error "This file should only be included through ParticleInterpolators.hpp"
+#endif
+
+#ifndef PARTICLEINTERPOLATORS_IMPL_HPP_
+#define PARTICLEINTERPOLATORS_IMPL_HPP_
+
+#include "InterpolationQueryParticle.hpp"
+#include "LagrangeInterpolation.hpp"
+#include "StateVariables.hpp"
+#include "VariableType.hpp"
+
+// amrex includes
+
+#include <AMReX_AmrLevel.H>
+#include <AMReX_AmrParGDB.H>
+#include <AMReX_ParallelDescriptor.H>
+#include <AMReX_ParmParse.H>
+
+template <int num_components>
+ParticleInterpolators<num_components>::ParticleInterpolators(
+    const BoundaryConditions::params_t &a_bc_params, int a_start_comp,
+    bool a_verbosity)
+    : m_gr_amr(nullptr), m_initialized(false), m_start_comp(a_start_comp),
+      m_bc_params(a_bc_params), m_verbosity(a_verbosity)
+{
+}
+
+// initialise everything and perform some sanity checks
+template <int num_components>
+void ParticleInterpolators<num_components>::set_gramr_ptr(GRAMR *gr_amr_ptr)
+{
+    // is GRAMR properly set?
+    AMREX_ASSERT(gr_amr_ptr != nullptr);
+    m_gr_amr = gr_amr_ptr;
+
+    this->Define(dynamic_cast<amrex::ParGDBBase *>(m_gr_amr->GetParGDB()));
+    m_initialized = true;
+
+    AMREX_ALWAYS_ASSERT(num_components >= 1);
+    AMREX_ALWAYS_ASSERT(m_start_comp >= 0);
+
+    // read in the physical bounds for reflective BC checks (it is sufficient to
+    // do this on lev = 0)
+    const amrex::Geometry &geom0 = m_gr_amr->getLevel(0).Geom();
+    const auto plo               = geom0.ProbLoArray();
+    const auto phi               = geom0.ProbHiArray();
+    for (int d = 0; d < AMREX_SPACEDIM; ++d)
+    {
+        m_prob_lo[d] = plo[d];
+        m_prob_hi[d] = phi[d];
+    }
```

Comment:
The spacing will be wrong in this suggestion.
```suggestion
    m_prob_lo               = geom0.ProbLoArray();
    m_prob_hi               = geom0.ProbHiArray();
```